### PR TITLE
Add RegisterWithEmailBloc for email registration flow

### PR DESCRIPTION
**Pull request by hyperdanyel**
# Add RegisterWithEmailBloc for Email Registration Flow

This PR implements the RegisterWithEmailBloc to handle the email registration flow with the following features:

- Email availability checking with debounce to prevent excessive API calls
- OTP (One-Time Password) sending functionality for email verification
- Support for editing email during registration process

The implementation includes:
- A new Bloc with appropriate events and states to manage the registration flow
- Email validation with real-time feedback on availability
- Integration with existing authentication use cases
- Comprehensive unit tests covering all possible states and edge cases

The Bloc handles three main events:
1. `RegisterWithEmailEmailChanged` - Checks if an email is already registered
2. `RegisterWithEmailContinuePressed` - Sends an OTP to the provided email
3. `RegisterWithEmailEditEmail` - Allows users to edit their email

This completes a key part of the authentication flow, enabling users to register with email verification.

### CODE REVIEW

**Comment by hyperdanyel** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/55?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#55** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/55?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/55?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#54** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/54?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#53** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/53?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#52** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/52?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#51** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/51?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#50** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/50?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#49** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/49?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#48** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/48?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#47** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/47?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/46?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#45** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/45?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#44** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/44?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/43?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#42** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/42?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#41** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/41?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#40** <a href="https://app.graphite.dev/github/pr/ripplearc/construculator-app/40?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/features/auth/presentation/bloc/register_with_email_bloc/register_with_email_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import 'package:equatable/equatable.dart';
+import 'package:construculator/features/auth/domain/usecases/check_email_availability_usecase.dart';
+import 'package:construculator/features/auth/domain/usecases/send_otp_usecase.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:rxdart/rxdart.dart';
+
+part 'register_with_email_event.dart';
+part 'register_with_email_state.dart';
+
+class RegisterWithEmailBloc
```
Comment (ripplearcgit) [inline lib/features/auth/presentation/bloc/register_with_email_bloc/register_with_email_bloc.dart:12]: Doc please
I assume we are missing docs for the all the blocs, events and states.
I am going stop commenting here, and let you address them. 
Thanks.

File: test/units/bloc/auth/register_with_email_bloc_test.dart [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+              'email': testEmail,
+              'created_at': DateTime.now().toIso8601String(),
+            },
+          ]);
+          bloc.add(RegisterWithEmailEmailChanged(testEmail));
+          await expectLater(
+            bloc.stream,
+            emitsInOrder([
+              RegisterWithEmailEmailCheckLoading(),
+              RegisterWithEmailEmailCheckSuccess(isEmailRegistered: true),
```
Comment (ripplearcgit) [inline test/units/bloc/auth/register_with_email_bloc_test.dart:70]: Instead of Success, should we used `Completed` to avoid confusion?

File: test/units/bloc/auth/register_with_email_bloc_test.dart [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        'emits [RegisterWithEmailOtpSendingLoading, RegisterWithEmailOtpSendingSuccess] when OTP sending succeeds',
+        build: () {
+          fakeSupabase.shouldThrowOnOtp = false;
+          return bloc;
+        },
+        act: (bloc) => bloc.add(RegisterWithEmailContinuePressed(testEmail)),
+        expect:
+            () => [
+              RegisterWithEmailOtpSendingLoading(),
+              RegisterWithEmailOtpSendingSuccess(),
```
Comment (ripplearcgit) [inline test/units/bloc/auth/register_with_email_bloc_test.dart:103]: Any chance for sending failure?

**Review by ripplearcgit — CHANGES_REQUESTED** [review]
Doc and mutation file please

File: lib/features/auth/presentation/bloc/register_with_email_bloc/register_with_email_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    required CheckEmailAvailabilityUseCase checkEmailAvailabilityUseCase,
+    required SendOtpUseCase sendOtpUseCase,
+  }) : _checkEmailAvailabilityUseCase = checkEmailAvailabilityUseCase,
+       _sendOtpUseCase = sendOtpUseCase,
+       super(RegisterWithEmailInitial()) {
+    on<RegisterWithEmailEmailChanged>(
+      _onEmailChanged,
+      transformer: (events, mapper) {
+        return events
+            .debounceTime(const Duration(milliseconds: 300))
```
[the remark this answers is not in the document]
Comment (nyelnizy) [reply lib/features/auth/presentation/bloc/register_with_email_bloc/register_with_email_bloc.dart:27]: updated

**Review by ripplearcgit — APPROVED** [review]
Please adde the mutation file before merging in

**Comment by nyelnizy** [comment]
#RunCheck
